File: Task3A_B/task3a_b/src/main/java/com/cps2004/ShipmentItem.java
package com.cps2004;

import com.fasterxml.jackson.databind.node.ObjectNode;

public class ShipmentItem {
    private int id;
    private int quantity;
    private Product product;
    private Package package_;

    // Default constructor
    public ShipmentItem() {
        this.id = 0;
        this.quantity = 0;
        this.product = null;
        this.package_ = null;
    }

    // Parameterized constructor
    public ShipmentItem(int id, int quantity, Product product, Package package_) {
        this.id = id;
        this.quantity = quantity;
        this.product = product;
        this.package_ = package_;
    }

    // Getters and setters
    public int getId() { return id; }
    public int getQuantity() { return quantity; }
    public Product getProduct() { return product; }
    public Package getPackage() { return package_; }
    public void setId(int id) { this.id = id; }
    public void setQuantity(int quantity) { this.quantity = quantity; }
    public void setProduct(Product product) { this.product = product; }
    public void setPackage(Package package_) { this.package_ = package_; }

    // Display shipment item information
    public void display_shipment_item() {
        System.out.println("Shipment item ID: " + this.id);
        System.out.println("Shipment item quantity: " + this.quantity);
        System.out.println("Shipment item product: ");
        this.product.display_product_info();
        System.out.println("Shipment item package: ");
        this.package_.viewPackage();
    }

}

File: Task3C/shipmentitem.cpp
#include "shipmentitem.h"

// Default constructor
ShipmentItem::ShipmentItem() {
    this->id = 0;
    this->quantity = 0;
    this->product = NULL;
    this->package = NULL;
}

// Parameterised constructor
ShipmentItem::ShipmentItem(int id, int quantity, Product* product, Package* package) {
    this->id = id;
    this->quantity = quantity;
    this->product = product;
    this->package = package;
}

// Getters and setters
int ShipmentItem::getId() {
    return this->id;
}

int ShipmentItem::getQuantity() {
    return this->quantity;
}

Product* ShipmentItem::getProduct() {
    return this->product;
}

Package* ShipmentItem::getPackage() {
    return this->package;
}

void ShipmentItem::setId(int id) {
    this->id = id;
}

void ShipmentItem::setQuantity(int quantity) {
    this->quantity = quantity;
}

void ShipmentItem::setProduct(Product* product) {
    this->product = product;
}

void ShipmentItem::setPackage(Package* package) {
    this->package = package;
}

// Display shipment item information
void ShipmentItem::display_shipment_item() {
    cout << "Shipment Item ID: " << this->id << endl;
    cout << "Quantity: " << this->quantity << endl;
    cout << "Product: " << endl;
    this->product->display_product_info();
    cout << "Package: " << endl;
    this->package->viewPackage();
}

// Save shipment item information to file
void ShipmentItem::save(ostream& out) const {
    this->product->save(out);
    this->package->save(out);
    out << this->id << "," << this->quantity << "\n";
}

// Load shipment item information from file
void ShipmentItem::load(istream& in) {
    this->product->load(in);
    this->package->load(in);

    string line;
    getline(in, line);
    stringstream ss(line);

    string token;
    getline(ss, token, ',');
    this->id = stoi(token);

    getline(ss, token, ',');
    this->quantity = stoi(token);
}

// Destructor
ShipmentItem::~ShipmentItem() {
    delete this->product;
    delete this->package;
}